dashboard/src/shared/schema.ts

```diff
@@ -1,17 +1,110 @@
+// WARN: yaml doesn't have updated definitions for TypeScript
+import * as YAML from "yaml";
 import { IBasicFormParam } from "./types";
 
-export function retrieveBasicFormParams(schema: any) {
-  // TBD
-  return {
-    username: {
-      name: "username",
-      path: "wordpressUsername",
-      value: "user",
-    } as IBasicFormParam,
-  };
+// Avoid to explicitly add "null" when an element is not defined
+import { nullOptions } from "yaml/types";
+nullOptions.nullStr = "";
+
+// retrieveBasicFormParams iterates over a JSON Schema properties looking for `form` keys
+// It uses the raw yaml to setup default values.
+// It returns a key:value map for easier handling.
+export function retrieveBasicFormParams(
+  defaultValues: string,
+  schema: any,
+): { [key: string]: IBasicFormParam } {
+  const basicFormParameters = {};
+  if (schema && schema.properties) {
+    const params = lookForFormParams(defaultValues, schema.properties);
+    if (params) {
+      params.forEach(p => (basicFormParameters[p.name] = p));
+    }
+  }
+  return basicFormParameters;
+}
+
+// getDocumentElem returns the YAML element given a root element
+export function getDocumentElem(rootElem: any, path: string): any | undefined {
+  // Split the path per keys
+  const splittedPath = path.split(".");
+  if (splittedPath.length === 1) {
+    // If the path doesn't contain inner objects, simply return
+    return rootElem.get(path);
+  }
+  // If the path contains inner elements, iterate until finding the last one
+  let res = rootElem.get(splittedPath[0]);
+  splittedPath.slice(1, splittedPath.length).forEach(p => (res = res && res.get(p)));
+  return res;
 }
 
+// setValue modifies the current values (text) based on a path
 export function setValue(values: string, path: string, newValue: any) {
-  // TBD
-  return values;
+  const doc = YAML.parseDocument(values);
+  if (path.includes(".")) {
+    // If the path points to an inner element we need to parse it
+    // and obtain the previous element to set it
+    // e.g.
+    // path = "a.b.c"
+    // itemToSet = "c"
+    // elem = "a.b"
+    // elem.set(itemToSet, value)
+    const splittedPath = path.split(".");
+    const itemToSet = splittedPath[splittedPath.length - 1];
+    splittedPath.pop();
+    const pathToSet = splittedPath.join(".");
+    const elem = getDocumentElem(doc, pathToSet);
+    if (!elem) {
+      // TODO: It should be possible to handle this situation and add a new object
+      throw new Error(`Unable to set the value ${path}. Parent element not found`);
+    }
+    elem.set(itemToSet, newValue);
+  } else {
+    doc.set(path, newValue);
+  }
+  return doc.toString();
+}
+
+// getDefaultValues returns the current value of an object based on YAML text and its path
+export function getDefaultValue(defaultValues: string, path: string): string | undefined {
```

Comment:
As above, this seems to be just a wrapper around `getDocumentElem` that returns the result as a string? It's not clear to me why we'd build the conversion into the function... worth checking if we can just use lodash here (or whether this extra function is needed at all.

Note, it doesn't seem to be related to `defaults` at all?

Reply:
the `Default` naming is confusing, yes